### PR TITLE
Add base test for extensionarray setitem #23300

### LINKED ISSUE #23300: Add base test for ExtensionArray.__setitem__
The tests in https://github.com/pandas-dev/pandas/blob/master/pandas/tests/extension/base/setitem.py seem to mainly operate on ExtensionArrays that are boxed in a Series.

It would be good to directly test ExtensionArray.__setitem__.

Some of these tests could be parametrized with something like

```python
@pytest.fixture(params=[True, False])
def box(request):
    """Whether to box the data in a Series"""
    return request.param


def test_setitem_scalar_series(self, data, box):
    if box:
        data = pd.Series(data)
    data[0] = data[1
    assert arr[0] == data[1]
```

others may have to be written separately.

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [x] closes #23300 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/extension/base/setitem.py [hunk 1]
```diff
@@ -36,25 +38,27 @@ def test_setitem_sequence_mismatched_length_raises(self, data, as_array):
...
+            data = pd.Series(data)
+        data[[0, 1]] = data[2]
+        assert data[0] == data[2]
+        assert data[1] == data[2]
 
     @pytest.mark.parametrize('setter', ['loc', 'iloc'])
-    def test_setitem_scalar(self, data, setter):
-        arr = pd.Series(data)
-        setter = getattr(arr, setter)
+    def test_setitem_scalar(self, data, box, setter):
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/setitem.py]: I think this will not work, since `loc` and `iloc` are only available on Series.

Comment (charlesdong1991) [reply pandas/tests/extension/base/setitem.py]: changed.
thanks for the quick feedback!

File: pandas/tests/extension/base/setitem.py [hunk 2]
```diff
@@ -36,25 +38,27 @@ def test_setitem_sequence_mismatched_length_raises(self, data, as_array):
         with tm.assert_raises_regex(ValueError, xpr.format('slice')):
             ser[slice(3)] = value
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/tests/extension/base/setitem.py]: you mean something like `assert ser[[0, 1]] == value`? sorry, i am a newbie, don't know if i understand correctly.

**Comment by TomAugspurger** [comment]
Looks like the fixture name `box` is already taken

```
    raise ValueError("duplicate %r" % (arg,))
E   ValueError: duplicate 'box'
--------------------- generated xml file: /tmp/single.xml ----------------------
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
============ 7 skipped, 35608 deselected, 1 error in 35.79 seconds =============

```

You could rename it to `box_in_series` perhaps.


FYI, you may want to run the tests locally for a faster feedback loop. `pytest pandas/tests/extension` or even just `pytest pandas/tests/extension/decimal`.

File: pandas/conftest.py [hunk 3]
```diff
@@ -501,6 +501,12 @@ def mock():
         return pytest.importorskip("mock")
 
 
+@pytest.fixture(params=[True, False])
+def box_in_series(request):
```
Comment (jreback) [inline pandas/conftest.py]: put this in the extension conftest

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/conftest.py (hunk 3 above)
Comment (charlesdong1991) [reply pandas/conftest.py]: done, thanks!

File: pandas/tests/extension/base/setitem.py [hunk 4]
```diff
@@ -32,22 +34,25 @@ def test_setitem_sequence_mismatched_length_raises(self, data, as_array):
...
-    def test_setitem_sequence_broadcasts(self, data):
-        arr = pd.Series(data)
-
-        arr[[0, 1]] = data[2]
-        assert arr[0] == data[2]
-        assert arr[1] == data[2]
+        assert ser._values[slice(3)] == value
+
+    def test_setitem_empty_indxer(self, data, box_in_series):
+        if box_in_series:
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/setitem.py]: The indentation is messed up here. This test is basically skipped when box_in_series is False.

You should dedent on line 47 and change the assert to `self.assert_equal`.

File: pandas/tests/extension/base/setitem.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/tests/extension/base/setitem.py]: Can you add these asserts?

File: pandas/tests/extension/base/setitem.py (hunk 4 above)
Comment (charlesdong1991) [reply pandas/tests/extension/base/setitem.py]: changed... appreciate!! and happy to learn a lot about tests... ^^

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks good.

@TomAugspurger IIRC on Sparse you were showing a warning on setitem? is this tested ?

**Comment by TomAugspurger** [comment]
SparseArray doesn't implement `__setitem__` (old one didn't either, though there's an open PR to implement it).

**Comment by TomAugspurger** [comment]
Thanks @charlesdong1991!

File: pandas/tests/extension/base/setitem.py [hunk 5]
```diff
@@ -32,22 +34,25 @@ def test_setitem_sequence_mismatched_length_raises(self, data, as_array):
         xpr = 'cannot set using a {} indexer with a different length'
         with tm.assert_raises_regex(ValueError, xpr.format('list-like')):
             ser[[0, 1]] = value
+            assert ser._values[[0, 1]] == value
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/setitem.py]: @TomAugspurger how will this assert be reached if the previous line raises an error?

Comment (TomAugspurger) [reply pandas/tests/extension/base/setitem.py]: Ah, sorry I missed that.

**Comment by charlesdong1991** [comment]
ah, sorry, i see, thanks for pointing it out... can PR be reopened?

Tom Augspurger <notifications@github.com> 于2018年10月26日周五 下午1:11写道：

**Comment by TomAugspurger** [comment]
No worries. I made a new PR to fix it.

**Pull request merged**